libdrgn/python/symbol.c

```diff
@@ -3,23 +3,83 @@
 
 #include <inttypes.h>
 
+#include "drgn.h"
 #include "drgnpy.h"
+#include "symbol.h"
 
 PyObject *Symbol_wrap(struct drgn_symbol *sym, Program *prog)
 {
-	Symbol *ret = call_tp_alloc(Symbol);
+	Symbol *ret;
+
+	if (!sym->name_owned && !prog) {
+		/* If a symbol does not own its name, then it implicitly
+		 * shares the lifetime of its associate program. So, we
+		 * must have it here. Symbols created via Python (Symbol_new())
+		 * always own their names. */
+		PyErr_SetString(PyExc_ValueError,
+				"Creating symbol with unowned name,"
+				" without program reference");
+		return NULL;
+	}
+	ret = call_tp_alloc(Symbol);
 	if (ret) {
 		ret->sym = sym;
 		ret->prog = prog;
-		Py_INCREF(prog);
+		if (prog)
+			Py_INCREF(prog);
 	}
 	return (PyObject *)ret;
 }
 
+PyObject *Symbol_new(PyTypeObject *subtype, PyObject *args, PyObject *kwds)
+{
+	struct drgn_symbol *sym;
+	static char *keywords[] = {"name", "address", "size", "binding", "kind", NULL};
+	const char *name;
+	struct index_arg address = {}, size = {};
+	struct enum_arg binding = {
+		.type = SymbolBinding_class,
+	};
+	struct enum_arg kind = {
+		.type = SymbolKind_class,
+	};
+
+	sym = malloc(sizeof(*sym));
+	if (!sym)
+		return set_drgn_error(&drgn_enomem);
+
+	if (!PyArg_ParseTupleAndKeywords(args, kwds, "sO&O&O&O&:Symbol", keywords,
+					 &name,
+					 index_converter, &address,
+					 index_converter, &size,
+					 enum_converter, &binding,
+					 enum_converter, &kind)) {
+		free(sym);
+		return NULL;
+	}
+
+	sym->name_owned = true;
+	sym->name = strdup(name);
```

Comment:
(This is a higher-level comment that may obsolete some of the other comments I left here.)

Instead of `strdup`ing the name, I think we can hold a reference to the original name `PyObject`. The reference we hold on `Symbol::prog` is only used to keep the name alive; we could replace that with a more general `PyObject *` that we need to hold for the lifetime of the symbol. Something along these lines:

```diff
diff --git a/libdrgn/python/drgnpy.h b/libdrgn/python/drgnpy.h
index 6f5fc6d5..65284838 100644
--- a/libdrgn/python/drgnpy.h
+++ b/libdrgn/python/drgnpy.h
@@ -173,7 +173,8 @@ typedef struct {
 
 typedef struct {
 	PyObject_HEAD
-	Program *prog;
+	// Reference we need to hold to keep the symbol name alive.
+	PyObject *name_obj;
 	struct drgn_symbol *sym;
 } Symbol;
 
@@ -288,7 +289,7 @@ Program *program_from_core_dump(PyObject *self, PyObject *args, PyObject *kwds);
 Program *program_from_kernel(PyObject *self);
 Program *program_from_pid(PyObject *self, PyObject *args, PyObject *kwds);
 
-PyObject *Symbol_wrap(struct drgn_symbol *sym, Program *prog);
+PyObject *Symbol_wrap(struct drgn_symbol *sym, PyObject *prog);
 
 PyObject *Thread_wrap(struct drgn_thread *drgn_thread);
 
diff --git a/libdrgn/python/program.c b/libdrgn/python/program.c
index c2103f48..a43ec63a 100644
--- a/libdrgn/python/program.c
+++ b/libdrgn/python/program.c
@@ -1012,7 +1012,7 @@ static PyObject *Program_symbols(Program *self, PyObject *args)
 		return NULL;
 	}
 	for (size_t i = 0; i < count; i++) {
-		PyObject *pysym = Symbol_wrap(symbols[i], self);
+		PyObject *pysym = Symbol_wrap(symbols[i], (PyObject *)self);
 		if (!pysym) {
 			/* Free symbols which aren't yet added to list. */
 			drgn_symbols_destroy(symbols, count);
@@ -1048,7 +1048,7 @@ static PyObject *Program_symbol(Program *self, PyObject *arg)
 	}
 	if (err)
 		return set_drgn_error(err);
-	ret = Symbol_wrap(sym, self);
+	ret = Symbol_wrap(sym, (PyObject *)self);
 	if (!ret) {
 		drgn_symbol_destroy(sym);
 		return NULL;
diff --git a/libdrgn/python/stack_trace.c b/libdrgn/python/stack_trace.c
index d738da14..125cc984 100644
--- a/libdrgn/python/stack_trace.c
+++ b/libdrgn/python/stack_trace.c
@@ -218,7 +218,7 @@ static PyObject *StackFrame_symbol(StackFrame *self)
 	err = drgn_stack_frame_symbol(self->trace->trace, self->i, &sym);
 	if (err)
 		return set_drgn_error(err);
-	PyObject *ret = Symbol_wrap(sym, prog);
+	PyObject *ret = Symbol_wrap(sym, (PyObject *)prog);
 	if (!ret) {
 		drgn_symbol_destroy(sym);
 		return NULL;
diff --git a/libdrgn/python/symbol.c b/libdrgn/python/symbol.c
index 0ea395e2..a625c295 100644
--- a/libdrgn/python/symbol.c
+++ b/libdrgn/python/symbol.c
@@ -7,26 +7,13 @@
 #include "drgnpy.h"
 #include "symbol.h"
 
-PyObject *Symbol_wrap(struct drgn_symbol *sym, Program *prog)
+PyObject *Symbol_wrap(struct drgn_symbol *sym, PyObject *name_obj)
 {
-	Symbol *ret;
-
-	if (!sym->name_owned && !prog) {
-		/* If a symbol does not own its name, then it implicitly
-		 * shares the lifetime of its associate program. So, we
-		 * must have it here. Symbols created via Python (Symbol_new())
-		 * always own their names. */
-		PyErr_SetString(PyExc_ValueError,
-				"Creating symbol with unowned name,"
-				" without program reference");
-		return NULL;
-	}
-	ret = call_tp_alloc(Symbol);
+	Symbol *ret = call_tp_alloc(Symbol);
 	if (ret) {
 		ret->sym = sym;
-		ret->prog = prog;
-		if (prog)
-			Py_INCREF(prog);
+		ret->name_obj = name_obj;
+		Py_INCREF(name_obj);
 	}
 	return (PyObject *)ret;
 }
@@ -35,7 +22,7 @@ PyObject *Symbol_new(PyTypeObject *subtype, PyObject *args, PyObject *kwds)
 {
 	struct drgn_symbol *sym;
 	static char *keywords[] = {"name", "address", "size", "binding", "kind", NULL};
-	const char *name;
+	PyObject *name_obj;
 	struct index_arg address = {}, size = {};
 	struct enum_arg binding = {
 		.type = SymbolBinding_class,
@@ -48,8 +35,8 @@ PyObject *Symbol_new(PyTypeObject *subtype, PyObject *args, PyObject *kwds)
 	if (!sym)
 		return set_drgn_error(&drgn_enomem);
 
-	if (!PyArg_ParseTupleAndKeywords(args, kwds, "sO&O&O&O&:Symbol", keywords,
-					 &name,
+	if (!PyArg_ParseTupleAndKeywords(args, kwds, "O!O&O&O&O&:Symbol",
+					 keywords, &PyUnicode_Type, &name_obj,
 					 index_converter, &address,
 					 index_converter, &size,
 					 enum_converter, &binding,
@@ -57,18 +44,17 @@ PyObject *Symbol_new(PyTypeObject *subtype, PyObject *args, PyObject *kwds)
 		free(sym);
 		return NULL;
 	}
+	const char *name = PyUnicode_AsUTF8(name_obj);
+	if (!name)
+		return NULL;
 
-	sym->name_owned = true;
-	sym->name = strdup(name);
-	if (!sym->name) {
-		free(sym);
-		return set_drgn_error(&drgn_enomem);
-	}
+	sym->name_owned = false;
+	sym->name = name;
 	sym->address = address.uvalue;
 	sym->size = size.uvalue;
 	sym->binding = binding.value;
 	sym->kind = kind.value;
-	PyObject *ret = Symbol_wrap(sym, NULL);
+	PyObject *ret = Symbol_wrap(sym, name_obj);
 	if (!ret) {
 		drgn_symbol_destroy(sym);
 	}
@@ -78,8 +64,7 @@ PyObject *Symbol_new(PyTypeObject *subtype, PyObject *args, PyObject *kwds)
 static void Symbol_dealloc(Symbol *self)
 {
 	drgn_symbol_destroy(self->sym);
-	if (self->prog)
-		Py_DECREF(self->prog);
+	Py_XDECREF(self->name_obj);
 	Py_TYPE(self)->tp_free((PyObject *)self);
 }
 
```

This name is not owned by the `drgn_symbol` or the `drgn_program`, so either we need a third state or `drgn_symbol_copy` needs to always copy the name.

Reply:
This is a great suggestion, thanks!

Sadly the string situation is all sorts of messed up still, even with this improvement. For symbols which are created internally to Drgn via C code, we can usually guarantee the lifetime of the string to be the same as the lifetime as the program, so the string copying situation looks like this:

1. Usually no copy on creation, due to the lifetime of the underlying symbol data source.
2. When you provide the symbol via the symbol finder API, no copy. (one malloc for the `drgn_symbol`)
3. When the corresponding Python Symbol object is created, it takes ownership of the `drgn_symbol` so no copy happens here, for the name or the struct.
4. When Python code accesses the name, we use `PyUnicode_FromString` which must make a copy, but that's truly unavoidable.

But if you're a Python Symbol finder:

1. One strdup on construction of the Symbol object, and we also malloc a new `drgn_symbol`. (Your improvement eliminates the strdup)
2. We use `drgn_symbol_copy()` and we `malloc()` a new `drgn_symbol` as well. We can't guarantee the lifetime of this name, so we'd need a strdup for the name. And you're right, we'd need a new state for the symbol to represent "don't free the name, but duplicate it when you copy it".
3. If the symbol is getting copied back out to a Python client, then `Symbol_wrap()` will end up taking ownership of the `drgn_symbol` and so no copy happens there. But the Python client would then copy the string on access to the name field, as above.

So for the first case, one copy of the string element itself, one malloc of the drgn_symbol. In the second case, the current branch would result in the name field being copied 3 times in order to be used, and the drgn_symbol structure is allocated twice. With your improvement, the 3 name copies becomes 2, and we still have allocated the `drgn_symbol` twice. It's not great :(

I have some wild idea where we do something like this instead of your approach:
```
diff --git a/libdrgn/python/program.c b/libdrgn/python/program.c
index c2103f4..091251b 100644
--- a/libdrgn/python/program.c
+++ b/libdrgn/python/program.c
@@ -547,16 +547,21 @@ static struct drgn_error *py_symbol_find_fn(const char *name, uint64_t addr,
 			err = drgn_error_from_python();
 			goto out;
 		}
-		_cleanup_free_ struct drgn_symbol *sym = malloc(sizeof(*sym));
-		if (!sym) {
-			err = &drgn_enomem;
-			goto out;
-		}
-		err = drgn_symbol_copy(sym, ((Symbol *)item)->sym);
-		if (err)
-			goto out;
+		if (Py_REFCNT(item) == 1) {
+			err = drgn_symbol_result_builder_add(builder, ((Symbol *)item)->sym);
+			((Symbol *)item)->sym = NULL;
+		} else {
+			_cleanup_free_ struct drgn_symbol *sym = malloc(sizeof(*sym));
+			if (!sym) {
+				err = &drgn_enomem;
+				goto out;
+			}
+			err = drgn_symbol_copy(sym, ((Symbol *)item)->sym);
+			if (err)
+				goto out;
 
-		err = drgn_symbol_result_builder_add(builder, no_cleanup_ptr(sym));
+			err = drgn_symbol_result_builder_add(builder, no_cleanup_ptr(sym));
+		}
 	}
 
 out:
```

We have to copy the string on creation of the Python Symbol object, and allocate a `drgn_symbol`. But if the Python symbol finder function simply returns the symbol without retaining a reference, we can take ownership of it. We can avoid allocating a second `drgn_symbol` and avoid that copy. The result is that the name is copied twice, and `drgn_symbol` allocated once.

I don't know if this is better though. Just an alternative which avoids the third state for the `drgn_symbol` object, and reduces the amount we allocate a `drgn_symbol`.

Reply:
Your idea is clever, although I think the `Py_REFCNT(item) == 1` check isn't quite enough. Consider:

```python3
my_symbols = [Symbol(...)]
def my_symbol_finder(name, address, first):
    return my_symbols
prog.add_symbol_finder(my_symbol_finder)
prog.symbol(...)
print(my_symbols[0])
```

I.e., there's only one reference on the symbol, but it can still be accessed via the list. So you probably also need a `Py_REFCNT(list) == 1` check. At this point it feels like we're playing with fire a bit :)

I don't love my third state idea either, though, so I'll leave it up to you which way to go.